Userland/Libraries/LibWeb/HTML/FormAssociatedElement.cpp

```diff
@@ -152,61 +165,248 @@ void FormAssociatedElement::reset_form_owner()
+        if (m_selection_start > relevant_value_length)
+            m_selection_start = relevant_value_length;
+
+        // 2. If the end of the selection is now past the end of the relevant value, set it to the
+        //    end of the relevant value.
+        if (m_selection_end > relevant_value_length)
+            m_selection_end = relevant_value_length;
+
+        // 3. If the user agent does not support empty selection, and both the start and end of the
+        //    selection are now pointing to the end of the relevant value, then instead set the
+        //    element's text entry cursor position to the end of the relevant value, removing any
+        //    selection.
+        // NOTE: We support empty selections.
+        return;
+    }
+
+    // 2. Otherwise, the element must have a text entry cursor position position. If it is now past
+    //    the end of the relevant value, set it to the end of the relevant value.
+    auto& document = form_associated_element_to_html_element().document();
+    auto const current_cursor_position = document.cursor_position();
+    if (current_cursor_position && text_node
+        && current_cursor_position->node() == text_node
+        && current_cursor_position->offset() > relevant_value_length) {
+        document.set_cursor_position(DOM::Position::create(document.realm(), *text_node, relevant_value_length));
+    }
+}
+
+// https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#dom-textarea/input-select
+WebIDL::ExceptionOr<void> FormAssociatedElement::select()
+{
+    // 1. If this element is an input element, and either select() does not apply to this element
+    //    or the corresponding control has no selectable text, return.
+    auto& html_element = form_associated_element_to_html_element();
+    if (is<HTMLInputElement>(html_element)) {
+        auto& input_element = static_cast<HTMLInputElement&>(html_element);
+        // FIXME: implement "or the corresponding control has no selectable text"
+        if (!input_element.select_applies())
+            return {};
+    }
+
+    // 2. Set the selection range with 0 and infinity.
+    return set_selection_range(0, NumericLimits<WebIDL::UnsignedLong>::max(), {});
+}
+
 // https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#dom-textarea/input-selectionstart
-WebIDL::UnsignedLong FormAssociatedElement::selection_start() const
+Optional<WebIDL::UnsignedLong> FormAssociatedElement::selection_start() const
 {
     // 1. If this element is an input element, and selectionStart does not apply to this element, return null.
-    // NOTE: This is done by HTMLInputElement before calling this function
+    auto const& html_element = form_associated_element_to_html_element();
+    if (is<HTMLInputElement>(html_element)) {
+        auto const& input_element = static_cast<HTMLInputElement const&>(html_element);
+        if (!input_element.selection_or_range_applies())
+            return {};
+    }
```

Comment:
Hmm was there a reason to change the semantics here? Seems like letting HTMLInputElement "override" (so-to-speak) the bindings methods to handle its special step was a bit nicer. Is the intent that something other than bindings will invoke these methods?

Reply:
This specification is defined in one place and is used by both HTMLInputElement and HTMLTextAreaElement. I could argue that the code follows the spec more closely now, because it now no longer needs to be split up between a base method and an element-specific override or binding.

Reply:
Marking this chain resolved after https://github.com/LadybirdBrowser/ladybird/pull/1193#discussion_r1731978683

We can always iterate later if we come up with something else :)